app/src/main/java/ch/epfl/favo/common/DatabaseWrapper.java

```diff
@@ -45,17 +53,50 @@ public static String generateRandomId() {
     return sb.toString();
   }
 
-  public static void addDocument(String key, Map document, String collection)
+  public static void addDocument(String key, Map document, String collection) {
+    new TaskToFutureAdapter<>(getCollectionReference(collection).document(key).set(document));
+  }
+
+  public static void removeDocument(String key, String collection) {
+    Task<Void> deleteTask = getCollectionReference(collection).document(key).delete();
+    new TaskToFutureAdapter<>(deleteTask);
+  }
+
+  public static void updateDocument(String key, Map<String, Object> updates, String collection) {
+    Task<Void> updateTask = getCollectionReference(collection).document(key).update(updates);
+    new TaskToFutureAdapter<>(updateTask);
+  }
+
+  public static CompletableFuture<Map> getDocument(String key, String collection)
```

Comment:
I'd prefer to have the return type be just a `Map`. Is there any way to do this asynchronously?

Reply:
Or at least I'd like to abstract `CompleteableFuture` and callbacks away from the caller.

Reply:
No there isn't, but that's actually a good thing. As mentioned in an earlier post, you cannot (and shouldn't!) try to fight against the asynchronous constructs. They are here to help you and your app, so try to embrace them (I know it's not easy to wrap our head around the concept as first). If you prefer to use callbacks as parameters instead of Futures that's totally fine, if you feel more at ease with it. You want the caller to be aware that this is an event that you do not know when it will terminate in the future: for instance, your UI wants to be informed when the event is still pending (so that it can display a loading animation and do other tasks meanwhile) and when it finishes (to actually populate the data). The Future is a nice abstraction around this unknown time delay semantic.

Reply:
Thanks!